src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        match self {
+            Error::NotImplementedError(e)
+            | Error::ConfigError(e)
+            | Error::ApiError(e)
+            | Error::CallbackError(e) => {
+                write!(f, "{}", e)
+            }
+        }
+    }
+}
+
+/// EvidenceCreationCb is the function signature of the application callback.
+/// The application is passed the session nonce and the list of supported
+/// evidence media types and shall return the computed evidence together with
+/// the selected media type.
+type EvidenceCreationCb =
+    fn(nonce: &Vec<u8>, accepted: &Vec<String>) -> Result<(Vec<u8>, String), Error>;
+
+/// A builder for ChallengeResponse objects
+pub struct ChallengeResponseBuilder {
+    base_url: Option<String>,
+    http_client: Option<reqwest::blocking::Client>,
+    evidence_creation_cb: Option<EvidenceCreationCb>,
+}
+
+impl ChallengeResponseBuilder {
+    /// default constructor
+    pub fn new() -> Self {
+        Self {
+            base_url: None,
+            http_client: None,
+            evidence_creation_cb: None,
+        }
+    }
+
+    /// Use this method to supply the base URL of the service implementing the
+    /// challenge-response API.  E.g.,
+    /// "https://veraison.example/challenge-response/v1/".
+    pub fn with_base_url(mut self, v: String) -> ChallengeResponseBuilder {
+        self.base_url = Some(v);
+        self
+    }
+
+    /// Use this method to supply a fully configured reqwest::blocking::Client
+    /// used for communicating with the Veraison service.
+    pub fn with_http_client(mut self, v: reqwest::blocking::Client) -> ChallengeResponseBuilder {
+        self.http_client = Some(v);
+        self
+    }
+
+    /// Use this method to provide the application callback that is used to
+    /// build the attestation evidence for this verification session.  See
+    /// EvidenceCreationCb.
+    pub fn with_evidence_creation_cb(mut self, v: EvidenceCreationCb) -> ChallengeResponseBuilder {
+        self.evidence_creation_cb = Some(v);
+        self
+    }
+
+    /// Instantiate a valid ChallengeResponse object, or fail with an error.
+    pub fn build(self) -> Result<ChallengeResponse, Error> {
+        let base_url_str = self
+            .base_url
+            .ok_or(Error::ConfigError("missing API endpoint".to_string()))?;
+
+        Ok(ChallengeResponse {
+            base_url: url::Url::parse(&base_url_str)
+                .map_err(|e| Error::ConfigError(e.to_string()))?,
+            http_client: self
+                .http_client
+                .ok_or(Error::ConfigError("missing HTTP client".to_string()))?,
+            evidence_creation_cb: self.evidence_creation_cb.ok_or(Error::ConfigError(
+                "missing evidence creation callback".to_string(),
+            ))?,
+        })
+    }
+}
+
+/// The object on which one or more challenge-response verification sessions can
+/// be run.  Always use the ChallengeResponseBuilder to instantiate it.
```

Comment:
```suggestion
/// be run.  Always use the [ChallengeResponseBuilder] to instantiate it.
```

This will [create a link](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html) to the documentation of the builder.